### PR TITLE
Add force and field constraints

### LINKED ISSUE #793: Add equilibria constraints for field and force using the new API
## Description of issue / requirement to address

#530 brought in a new way of setting up optimisation problems and mixing in constraints into optimisation problems for equilibria.

## Proposed solution

We need to add constraint objects for the old force and field constraints.

## Alternative solutions
We could keep the old ones alive for a while so that we're always in a position to run BLUEPRINT for EU-DEMO. Hopefully that won't be necessary for much longer...

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Closes #793 

## Description
Just adds in some functions for force and field constraints a la EU-DEMO. Not tested or threaded in yet, thought I'd split this down into smaller PRs.

## Interface Changes
None yet

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Review by je-cook — APPROVED** [review]
Dunno if this is finished or not but this looks fine, but I think @alexanderianblair should take a look when he gets a chance

**Comment by CoronelBuendia** [comment]
> Dunno if this is finished or not but this looks fine, but I think @alexanderianblair should take a look when he gets a chance

Yeah I think it is, and yes would like Alex to check!

File: bluemira/equilibria/opt_constraints.py [hunk 1]
```diff
@@ -93,3 +95,115 @@ def current_midplane_constraint(constraint, vector, grad, eq, radius, inboard=Tr
...
+    """
+    # get coil force and jacobian
+    F, dF = eq.force_field.calc_force(vector * scale)  # noqa :N803
+    F /= scale  # Scale down to MN
+    # dF /= self.scale
+
+    # calculate constraint jacobian
+    if grad.size > 0:
+        # PFz lower bound
+        grad[:n_PF] = -dF[:n_PF, :, 1]
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/equilibria/opt_constraints.py]: Don't disagree; this is a bit of a halfway house to moving the optimisers across to the new structure. It's possible the `ForceField` object could be dissolved when we come to implement this in earnest. For now I'm just trying to set myself up with some tools to set up the old problems relatively easily from scratch.

File: bluemira/equilibria/opt_constraints.py [hunk 2]
```diff
@@ -93,3 +95,115 @@ def current_midplane_constraint(constraint, vector, grad, eq, radius, inboard=Tr
...
+            for j in range(n_CS - 1):  # evaluate each gap in CS stack
+                # CS separation constraint Jacobians
+                f_up = np.sum(dF[n_PF : n_PF + j + 1, :, 1], axis=0)
+                f_down = np.sum(dF[n_PF + j + 1 :, :, 1], axis=0)
+                grad[2 * n_PF + 2 + j] = f_up - f_down
+
+    # vertical force on PF coils
+    pf_fz = F[:n_PF, 1]
+    # PFz lower bound
+    constraint[:n_PF] = -PF_Fz_max - pf_fz
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/equilibria/opt_constraints.py]: yup, v nice

**Review by alexanderianblair — APPROVED** [review]
All good now the issues above have been addressed and disclaimers added.

**Pull request merged**